File: tile.openstreetmap.org/ruby/rules.rb
#!/usr/bin/ruby
require 'xml/libxml'

class RenderRules
	# Opens the xml file and reads it into an easier and simplified structure
	def initialize(xmlfile)
		@rules = Array.new 
		document = XML::Document.file(xmlfile)
		document.find("//ruleset/rule").each do |rule|
			currule = {}
			currule['conditions'] = {}
			currule['style'] = {}
			rule.find('condition').each do |condition|
				currule['conditions'][condition['k']] = condition['v']
			end
			rule.find('style').each do |s|
				currule['style']['colour'] = s['colour']
				currule['style']['casing'] = s['casing']
				currule['style']['width'] = s['width']
				currule['style']['dash'] = s['dash']
				currule['style']['z-index'] = s['z-index'].to_i
				currule['style']['image'] = s['image']
				currule['style']['text'] = s['text']
			end
			@rules.push(currule) 
		end
	end

	# TODO: OO this
	def get_style(supplied_keyvals)
		prevhits = 0
		style = {}
		@rules.each do |rule|
			hits = 0
			supplied_keyvals.each do |supplied_key,supplied_val|
				unless rule['conditions'][supplied_key] == nil
					if rule['conditions'][supplied_key] == supplied_val
						hits = hits + 1
					else
						hits = 0
						break
					end
				end
			end

			if hits > prevhits
				prevhits = hits
				style = rule['style']
			end
		end

		style['casing'] = 'black' if style['casing']==nil and 
					style['dash'] == nil
		#style['casing'] = nil if style['casing'] = 'none'
		style['colour'] = 'green' if style['colour']==nil
		style['width'] = "0,0,0,0,0,0,0,0,0,1,1,1,4,6,8" if style['width']==nil
		style['z-index'] = 0 if style['z-index']==nil

		return style
	end
end

File: freemap/rules.php
<?php
################################################################################
# This file forms part of the Freemap source code.                             #
# (c) 2004-06 Nick Whitelegg (Hogweed Software)                                #
# Licenced under the Lesser GNU General Public Licence; see COPYING            #
# for details.                                                                 #
################################################################################
require_once('defines.php');
//header("Content-type: text/xml");

#globals
$suppliedKeyvals = null;
$inDoc=$inRule=$inCondition=$inStyle=$inGpsmapcode=$inGpstype=false;
$curConditions=null;
$curRule=null;
$rules = array();
#end globals

function readStyleRules($rulefile)
{
	global $rules;

	$parser = xml_parser_create();
	xml_set_element_handler($parser,"on_start_element_rules",
				"on_end_element_rules");
	xml_set_character_data_handler($parser,"on_characters_rules");

	$fp = fopen($rulefile,"r");
	while($line=fread($fp,4096))
	{
		if(!xml_parse($parser,$line))
		{
			fclose($fp);
			return false;
		}
	}

	fclose($fp);
	xml_parser_free($parser);
	return $rules; 
}

#NB the PHP expat library reads in all tags as capitals - even if they're
#lower case!!!
function on_start_element_rules($parser,$element,$attrs)
{
	global $inDoc, $inRule, $curConditions, $inCondition, $curRule,
			$inStyle, $inGpsmapcode, $inGpstype;

	if($element=="RULESET")
	{
		$inDoc = true;
	}
	elseif($inDoc)
	{
		if($element=="RULE")
		{
			$inRule=true;
			$curRule = array();
		}
		elseif($element=="CONDITION" && $inRule)
		{
			$inCondition=true;
			if(isset($attrs["K"]) && isset($attrs["V"]))
				$curRule["conditions"][$attrs["K"]]=$attrs["V"];
		}
		elseif($element=="STYLE" && $inRule)
		{
			$inStyle=true;
			foreach($attrs as $name => $value)
				$curRule[strtolower($name)] = $value;
		}
		elseif($element=="GPSMAPCODE" && $inRule)
			$inGpsmapcode = true;
		elseif($element=="GPSTYPE" && $inRule)
			$inGpstype = true;
	}
}

function on_end_element_rules($parser,$element)
{
	global $inDoc, $inRule, $curConditions, $inCondition, $curRule,
			$inStyle, $rules, $inGpsmapcode, $inGpstype;

	if($element=="RULESET")
		$inDoc = false;
	elseif($inDoc && $element=="RULE")
	{
		$inRule = false;
		$rules[] = $curRule;
	}
	elseif($inRule && $element=="CONDITION")
		$inCondition = false;
	elseif($inRule && $element=="STYLE")
		$inStyle = false;
	elseif($element=="GPSMAPCODE" && $inRule)
		$inGpsmapcode = false;
	elseif($element=="GPSTYPE" && $inRule)
		$inGpstype = false;
}

function on_characters_rules($parser, $characters)
{
	global $curRule, $inGpsmapcode, $inGpstype;
	if($inGpsmapcode)
		$curRule["gpsmapcode"] = $characters;
	elseif($inGpstype)
		$curRule["gpstype"] = $characters;
}

function getStyle($rules,$suppliedKeyvals)
{
	$prevHit = 0;
	$style=array();
	// defaults
	$style["colour"] = "220,220,220";
	$style["width"] = 1;

	foreach($rules as $curRule)
	{
		$hit = 0;

		// Go through all supplied Keyvals
		foreach($suppliedKeyvals as $k=>$v)
		{
			// If the current ruleset includes the current key...
			if(isset($curRule["conditions"][$k]))
			{
				// If key and value match, increase no. of hits
				if($curRule["conditions"][$k]==$v)
				{
					$hit++;
				}
				// If values are different, we definitely don't want this rule
				else
				{
					$hit=0;
					break;
				}
			}
		}
		// If more rule "hits" this time, this is seen as greater specificity,
		// so take on this rule
		if($hit > $prevHit)
		{
			$style=array();
			foreach($curRule as $k=>$v)
				if($k!="conditions")
					$style[$k] = $v; 
			$prevHit = $hit;
		}
	}

	/*
	if(!isset($style["dash"]) && !isset($style["casing"]))
		$style["casing"] = "0,0,0";
	*/

	return $style;
}

?>
